src/Host/Client/Impl/Host/LocalRHostConnector.cs

```diff
@@ -31,134 +33,150 @@ internal sealed class LocalRHostConnector : IRHostConnector {
-                let pos = rnd.NextDouble()
-                orderby pos
-                select port;
-
-            var transportTcs = new TaskCompletionSource<IMessageTransport>();
-            foreach (var port in ports) {
-                cancellationToken.ThrowIfCancellationRequested();
-
-                server = new WebSocketServer(port) {
-                    ReuseAddress = false,
-                    WaitTime = HeartbeatTimeout,
-                };
-
-                server.AddWebSocketService("/", () => CreateWebSocketMessageTransport(transportTcs));
-
-                try {
-                    server.Start();
-                    break;
-                } catch (SocketException ex) {
-                    if (ex.SocketErrorCode == SocketError.AddressAlreadyInUse) {
-                        server = null;
-                    } else {
-                        throw new MessageTransportException(ex);
+            base.Dispose();
+
+            if (_brokerProcess != null) {
+                if (!_brokerProcess.HasExited) {
+                    try {
+                        _brokerProcess.Kill();
+                    } catch (Win32Exception) {
+                    } catch (InvalidOperationException) {
                     }
-                } catch (WebSocketException ex) {
-                    throw new MessageTransportException(ex);
-                }
-            }
 
-            if (server == null) {
-                throw new MessageTransportException(new SocketException((int)SocketError.AddressAlreadyInUse));
+                    _brokerProcess = null;
+                }
             }
+        }
 
-            var psi = new ProcessStartInfo {
-                FileName = rhostExe,
-                UseShellExecute = false
+        protected override HttpClientHandler GetHttpClientHandler() {
+            return new HttpClientHandler {
+                Credentials = _credentials
             };
+        }
 
-            var shortHome = new StringBuilder(NativeMethods.MAX_PATH);
-            NativeMethods.GetShortPathName(_rHome, shortHome, shortHome.Capacity);
-            psi.EnvironmentVariables["R_HOME"] = shortHome.ToString();
+        protected override void ConfigureWebSocketRequest(HttpWebRequest request) {
+            request.AuthenticationLevel = AuthenticationLevel.MutualAuthRequested;
+            request.Credentials = _credentials;
+        }
 
-            psi.EnvironmentVariables["PATH"] = rBinPath + ";" + Environment.GetEnvironmentVariable("PATH");
+        protected override async Task ConnectToBrokerAsync() {
+            DisposableBag.ThrowIfDisposed();
+            await TaskUtilities.SwitchToBackgroundThread();
 
-            if (name != null) {
-                psi.Arguments += " --rhost-name " + name;
+            try {
+                if (await _connectLock.WaitAsync()) {
+                    if (!_isConnected) {
```

Comment:
Side note: this is basically best effort recovery in case broker crashes, or user kills it from Task Manager etc. There are going to be unavoidable races here, because if the process crashes, Exited event is delivered asynchronously, so in the worst case, we won't know until we send the HTTP request. I contemplated adding a loop around that whole thing so that if request fails, it checks if process is still alive and tries to restart it, but I'm not sure if it's a good idea - if the broker does crash repeatedly, we don't want to keep restarting it over and over again. 

---
In reply to: [75033980](https://github.com/Microsoft/RTVS/pull/2173#discussion_r75033980) [](ancestors = 75033980,75033083,75031624)